Brooks/LanguageRenderer/NewCSharpRenderer.hs

```diff
@@ -0,0 +1,561 @@
+    BooleanExpression(..), ValueExpression(..), Selector(..), FunctionSym(..), 
+    SelectorFunction(..), StatementSym(..), ControlStatementSym(..), ScopeSym(..),
+    MethodTypeSym(..), ParameterSym(..), MethodSym(..), StateVarSym(..), 
+    ClassSym(..), ModuleSym(..))
+import NewLanguageRenderer (Terminator(..), fileDoc', 
+    moduleDocD, classDocD, enumDocD,
+    enumElementsDocD, multiStateDocD, blockDocD, bodyDocD, outDocD,
+    printFileDocD, boolTypeDocD, 
+    intTypeDocD, charTypeDocD, stringTypeDocD, typeDocD, listTypeDocD, voidDocD,
+    constructDocD, stateParamDocD, paramListDocD, methodDocD, methodListDocD, 
+    stateVarDocD, stateVarListDocD, ifCondDocD, switchDocD, forDocD, 
+    forEachDocD, whileDocD, stratDocD, assignDocD, plusEqualsDocD, plusPlusDocD,
+    varDecDocD, varDecDefDocD, listDecDocD, listDecDefDocD, objDecDefDocD, 
+    constDecDefDocD, statementDocD, returnDocD,
+    commentDocD, notOpDocD, negateOpDocD, unOpDocD, equalOpDocD, 
+    notEqualOpDocD, greaterOpDocD, greaterEqualOpDocD, lessOpDocD, 
+    lessEqualOpDocD, plusOpDocD, minusOpDocD, multOpDocD, divideOpDocD, 
+    moduloOpDocD, andOpDocD, orOpDocD, binOpDocD, binOpDocD', litTrueD, litFalseD,
+    litCharD, litFloatD, litIntD, litStringD, defaultCharD, defaultFloatD, defaultIntD, 
+    defaultStringD, varDocD, extVarDocD, selfDocD, argDocD, enumElemDocD, objVarDocD, 
+    inlineIfDocD, funcAppDocD, extFuncAppDocD, stateObjDocD, listStateObjDocD, 
+    notNullDocD, listIndexExistsDocD, funcDocD, castDocD, listSetDocD, 
+    listAccessDocD, objAccessDocD, castObjDocD, breakDocD, continueDocD, 
+    staticDocD, dynamicDocD, privateDocD, publicDocD, dot, new, 
+    observerListName, doubleSlash, addCommentsDocD, callFuncParamList, 
+    getterName, setterName, setMain, setEmpty, statementsToStateVars)
+import Helpers (oneTab, tripFst, tripSnd, tripThird, liftA4, liftA5, liftA6, 
+  liftA7, liftList, lift1List, lift3Pair, lift4Pair, liftPairFst, liftTripFst)
+
+import Prelude hiding (break,print,(<>),sin,cos,tan,floor)
+import qualified Data.Map as Map (fromList,lookup)
+import Control.Applicative (Applicative, liftA2, liftA3)
+import Text.PrettyPrint.HughesPJ (Doc, text, (<>), (<+>), parens, comma, empty,
+  equals, semi, vcat, lbrace, rbrace, render, colon, render)
+
+newtype CSharpCode a = CSC {unCSC :: a}
+
+instance Functor CSharpCode where
+    fmap f (CSC x) = CSC (f x)
+
+instance Applicative CSharpCode where
+    pure = CSC
+    (CSC f) <*> (CSC x) = CSC (f x)
+
+instance Monad CSharpCode where
+    return = CSC
+    CSC x >>= f = f x
+
+instance PackageSym CSharpCode where
+    type Package CSharpCode = ([(Doc, Label, Bool)], Label)
+    packMods n ms = liftPairFst (sequence ms, n)
+
+instance RenderSym CSharpCode where
+    type RenderFile CSharpCode = (Doc, Label, Bool)
+    fileDoc code = liftTripFst (liftA3 fileDoc' (top code) (fmap tripFst code) bottom, tripSnd $ unCSC code, tripThird $ unCSC code)
```

Comment:
I'm thinking that using a record for `(Doc, Label, Bool)` would enable nicer names than `tripFst` et al, and thus make the code nicer.

Reply:
It pained me when I resorted to using triples and having to write `tripFst` et al. This was never intended to be a long-term solution. We've talked before about making our instances `Reader`s so that we don't have to explicitly pass around these `Label`s and `Bool`s like we do here. However, it will be easy to replace this with a record for now, so I'll do that.

Reply:
Added records and removed functions for working with triples in 77139c28eaa4cc4d2b155e47de2fcfe1c7682fbb and 4755f9b996c33bd70c0bd9b98f9236ff367a50d5.